### PR TITLE
[eval] Fix multi-processing bug (again^3) & allow set EXP_NAME for each `run_infer`

### PR DESCRIPTION
**Pull request by xingyaoww**
**Short description of the problem this fixes or functionality that this introduces. This may be used for the CHANGELOG**

Hopefully fix https://github.com/All-Hands-AI/OpenHands/pull/3875#issuecomment-2351838834.

(Can't count this is how many time we try to fix the eval bug 😭 )

---
**Give a summary of what the PR does, explaining any non-trivial design decisions**

- Revert the "re-add error instance to the queue" logic introduced in https://github.com/All-Hands-AI/OpenHands/pull/3854 to simplify things.
- Right now, we simply re-try in a wrapper for each `process_instance_func` - we keep running the `func` for 5 time straight until the model got it right.

---
**Link of any specific issues this addresses**

### CODE REVIEW

**Comment by xingyaoww** [comment]
cc @enyst .

I'm running eval now to see if this fixes stuff

**Comment by enyst** [comment]
Thanks for looking into it!
FWIW I ran into it again, but this time admittedly I stayed for a looong time on some debugger breakpoints and then it couldn't pick up. 🤷 I'm not blaming it - _this time._ 😅

**Comment by xingyaoww** [comment]
@enyst Just finish running a few hundred instances - it seems to work this time?

**Review by neubig — APPROVED** [review]
LGTM!

**Comment by enyst** [comment]
LGTM, lets try it!

File: evaluation/utils/shared.py [hunk 1]
```diff
@@ -235,65 +229,58 @@ def prepare_dataset(
...
-        + '[You may ignore this error if it is a transient issue - the instance will be automatically retried.]'
-        + '-' * 10
-        + '\n'
-    )
-    instance_queue.put(instance)
-    pbar.total += 1
-    pbar.refresh()
+    Retry an instance up to max_retries times if it fails (e.g., due to transient network/runtime issues).
+    """
+    for attempt in range(max_retries):
```
Comment (li-boxuan) [inline evaluation/utils/shared.py:258]: This seems wrong. If `max_retries` is zero, this code does nothing, which is obviously wrong.

```suggestion
    for attempt in range(max_retries + 1):
```

File: evaluation/utils/shared.py [hunk 2]
```diff
@@ -235,65 +229,58 @@ def prepare_dataset(
...
+        try:
+            result = process_instance_func(instance, metadata, use_mp)
+            return result
+        except Exception as e:
+            if attempt == max_retries - 1:
```
Comment (li-boxuan) [inline evaluation/utils/shared.py:263]: Following the last comment

```suggestion
            if attempt == max_retries:
```

File: evaluation/utils/shared.py (hunk 1 above)
Comment (xingyaoww) [reply evaluation/utils/shared.py]: Thanks for catching this!

**Pull request merged**